File: leetcode/0014/python3/main.py
# coding=utf-8

class Solution:
    def longestCommonPrefix(self, strs: 'List[str]') -> 'str':
        prefix = ""
        for x in zip(*strs):
            if all(item == x[0] for item in x):
                prefix += x[0]
            else:
                break
        return prefix

if __name__ == "__main__":
    s = Solution()
    assert s.longestCommonPrefix(["flower", "flow", "flight"]) == "fl"
    assert s.longestCommonPrefix(["flower", "flow", "flight"]) == "fl";
    assert s.longestCommonPrefix(["dog", "racecar", "car"]) == "";
    assert s.longestCommonPrefix(["dog", "racecar", ""]) == "";
    assert s.longestCommonPrefix(["", "racecar", "car"]) == "";
    assert s.longestCommonPrefix([]) == "";
    assert s.longestCommonPrefix([""]) == "";
    assert s.longestCommonPrefix(["", ""]) == "";

File: leetcode/0014/cpp/main.cpp
#include <cassert>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.empty())
        {
            return string();
        }

        if (strs.size() == 1)
        {
            return strs[0];
        }

        string prefix;

        uint32_t pos = 0;
        while (pos < strs[0].size())
        {
            char c = strs[0][pos];
            for (uint32_t si = 1; si < static_cast<uint32_t>(strs.size()); si++)
            {
                if (pos >= strs[si].size() || strs[si][pos] != c)
                {
                    return prefix;
                }
            }

            prefix += c;
            ++pos;
        }

        return prefix;
    }
};

int main()
{
    Solution s;

    assert(s.longestCommonPrefix(vector<string>{"flower", "flow", "flight"}) == "fl");
    assert(s.longestCommonPrefix(vector<string>{"dog", "racecar", "car"}) == "");
    assert(s.longestCommonPrefix(vector<string>{"dog", "racecar", ""}) == "");
    assert(s.longestCommonPrefix(vector<string>{"", "racecar", "car"}) == "");
    assert(s.longestCommonPrefix(vector<string>{}) == "");
    assert(s.longestCommonPrefix(vector<string>{""}) == "");
    assert(s.longestCommonPrefix(vector<string>{"", ""}) == "");

    return 0;
}
